src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBaker.java

```diff
@@ -76,7 +126,207 @@ public String getJobName()
+            }
+            taskCompound.setTag(TAG_PRODUCTS, productsTaskList);
+            tasksTagList.appendTag(taskCompound);
+        }
+        compound.setTag(TAG_TASKS, tasksTagList);
+
+        @NotNull final NBTTagList furnacesTagList = new NBTTagList();
+        for (@NotNull final Map.Entry<BlockPos, Product> entry : furnaces.entrySet())
+        {
+            @NotNull final NBTTagCompound furnaceCompound = new NBTTagCompound();
+            BlockPosUtil.writeToNBT(furnaceCompound, TAG_FURNACE_POS, entry.getKey());
+
+            if(entry.getValue() != null)
+            {
+                entry.getValue().writeToNBT(furnaceCompound);
+            }
+
+            furnacesTagList.appendTag(furnaceCompound);
+        }
+        compound.setTag(TAG_FURNACES, furnacesTagList);
+    }
+
+    @Override
+    public void readFromNBT(@NotNull final NBTTagCompound compound)
+    {
+        tasks.clear();
+        super.readFromNBT(compound);
+        final NBTTagList taskTagList = compound.getTagList(TAG_TASKS, Constants.NBT.TAG_COMPOUND);
+        for (int i = 0; i < taskTagList.tagCount(); ++i)
+        {
+            final NBTTagCompound taskCompound = taskTagList.getCompoundTagAt(i);
+            final ProductState state = ProductState.values()[taskCompound.getInteger(TAG_STATE)];
+            final List<Product> products = new ArrayList<>();
+
+            final NBTTagList productTagList = taskCompound.getTagList(TAG_PRODUCTS, Constants.NBT.TAG_COMPOUND);
+            for (int j = 0; j < productTagList.tagCount(); ++j)
+            {
+                final NBTTagCompound productCompound = taskTagList.getCompoundTagAt(i);
+                final Product product = Product.createFromNBT(productCompound);
+                products.add(product);
+            }
+
+            tasks.put(state, products);
+        }
+
+        final NBTTagList furnaceTagList = compound.getTagList(TAG_FURNACES, Constants.NBT.TAG_COMPOUND);
+        for (int i = 0; i < furnaceTagList.tagCount(); ++i)
+        {
+            final NBTTagCompound furnaceCompound = furnaceTagList.getCompoundTagAt(i);
+            final BlockPos pos = BlockPosUtil.readFromNBT(furnaceCompound, TAG_FURNACE_POS);
+            final Product product = Product.createFromNBT(furnaceCompound);
+            furnaces.put(pos, product);
+        }
+    }
+
+    @Override
+    public void registerBlockPosition(@NotNull final Block block, @NotNull final BlockPos pos)
+    {
+        super.registerBlockPosition(block, pos);
+        if (block instanceof BlockFurnace && !furnaces.containsKey(pos))
+        {
+            addToFurnaces(pos);
+        }
+        markDirty();
+    }
+
+    /**
+     * Add a task to the tasks list.
+     *
+     * @param state   the state of the task.
+     * @param product the regarding product.
+     */
+    public void addToTasks(final ProductState state, final Product product)
+    {
+        if (tasks.containsKey(state))
```

Comment:
Then I wouldn't even need the "state" entry.
I'm only adding the product in code at places where it shouldn't be possible to have changed.